[BUGFIX beta] Fix basic component and helper resolution in engines.

rwjblue opened the pull request:
Rebase and update of https://github.com/emberjs/ember.js/pull/14112 (to use `symbolTable` throughout) which is a rebase of https://github.com/emberjs/ember.js/pull/14055.

Contains the following fixes:
- Clone additional parent dependencies into engine instances. Specifically service:-glimmer-environment (only for glimmer) and renderer:-dom or renderer:-inert
- Ensure that Cache lookups are performed using the correct owner.
- Ensure that engines are destroyed before top-level views so that top-level
  views are not re-instantiated during engine teardown.

asakusuma committed:
[BUGFIX beta] Fix basic component and helper resolution in engines.

Ensure that Cache lookups are performed using the correct owner.

Ensure that engines are destroyed _before_ top-level views so that top-level
views are not re-instantiated during engine teardown.

dgeb committed:
[BUGFIX beta] Clone additional parent dependencies into engine instances.

An engine instance needs to share the following with its parent:

* `service:-glimmer-environment` (only for glimmer)
* `renderer:-dom` or `renderer:-inert`, depending on whether the environment
  is interactive

asakusuma commented:
Looks like [`Applications with autoboot set to false do not autoboot`](https://github.com/emberjs/ember.js/blob/eecc9f99b58c02c2dc076a9abd816e53671933a0/packages/ember-application/tests/system/visit_test.js#L80) is failing but only in Firefox and with feature flags off.

Serabe commented:
@asakusuma it seems to be failing only sometimes for some merges :$ 
![image](https://cloud.githubusercontent.com/assets/36221/18022181/0e842458-6bed-11e6-8346-b8e1ae2f9eab.png)

Commit:
Use the correct owner for each template lookup.

Ensure that the owner being used is not the environments default
`Owner`, but it is the owner that the template is being looked up by.

For late bound templates this means the components owner, and for normal
templates (looked up from registry) it is passed to the factories
`.create` method by the container itself.

rwjblue commented:
Yes, we have an intermittent issue with FF_Current.

rwjblue commented:
I updated the owner specific stuff to ensure that we do not use the environments owner (which in the case of engines would be the consuming apps not the engines).  Still need to add a couple tests around this (where the engine and the app both have a given component/helper but the "right" one is used in each context).
